pydantic/tools.py

```diff
@@ -0,0 +1,20 @@
+from functools import lru_cache
+from typing import Any, Type, TypeVar
+
+__all__ = ('parse_as_type',)
+
+
+@lru_cache(maxsize=None)
+def _get_parsing_type(type_: Any, source: str) -> Any:
+    from pydantic.main import create_model
+
+    type_name = getattr(type_, '__name__', str(type_))
+    return create_model(f'ParsingModel[{type_name}] (for {source})', obj=(type_, ...))
```

Comment:
is this name that isn't valid python going to cause confusion or anything to break?

Maybe better to remove ` (for {source})`? We could add `docstring` as an argument to `create_model` and add more information there?

Reply:
I don't think the naming will cause anything to break; the generated classes are really an implementation detail and are basically private to this function.

Currently, the only way their existence should even be publicly detectable is if you get a parsing error when calling the function, and even then it will only show up in the traceback where we print the model name.

This was originally added to make it easier to figure out where this `ParsingModel` came from -- if you just saw a validation error for, `ParsingModel[int]`, and hadn't defined `ParsingModel` in your code, I think it could be relatively painful to discover what was going on (and that the error was coming from the call to `parse_as_model`).

But now that I think about it, if your error came from a call to this , you'd ultimately get the `parse_as_model` call in the traceback, and that's probably good enough for discoverability.

------

Thinking about this more though, I think we should probably just make it possible to override the generated name (similar to the recent change to GenericModel), since I could imagine framework writers wanting a little more control over the generated error messages. 

I'll think about if there is an easy way to accomplish this.

Reply:
I think we should remove ` (for {source})` and instead allow people to set their own names if they wish?